packages/react-components/react-combobox/src/components/Combobox/useCombobox.tsx

```diff
@@ -18,22 +19,104 @@ import type { ComboboxProps, ComboboxState } from './Combobox.types';
  */
 export const useCombobox_unstable = (props: ComboboxProps, ref: React.Ref<HTMLInputElement>): ComboboxState => {
   const baseState = useComboboxBaseState(props);
+  const {
+    activeOption,
+    clearSelection,
+    getIndexOfId,
+    getOptionsMatchingValue,
+    selectOption,
+    selectedOptions,
+    setActiveOption,
+    setOpen,
+    setValue,
+    value,
+  } = baseState;
+  const { allowFreeform, multiselect } = props;
 
   const { primary: triggerNativeProps, root: rootNativeProps } = getPartitionedNativeProps({
     props,
     primarySlotTagName: 'input',
     excludedPropNames: ['children', 'size'],
   });
 
+  // set active option and selection based on typing
+  const getOptionFromInput = (inputValue: string): OptionValue | null => {
+    // if there are commas in the value string, take the text after the last comma
+    const searchString = inputValue.split(',').pop();
+
+    if (!searchString || searchString.trim().length === 0) {
+      return null;
+    }
+
+    const matcher = (optionValue: string) => optionValue.toLowerCase().indexOf(searchString.trim().toLowerCase()) === 0;
+    const matches = getOptionsMatchingValue(matcher);
+
+    // return first matching option after the current active option, looping back to the top
+    if (matches.length > 1 && activeOption) {
+      const startIndex = getIndexOfId(activeOption.id);
+      const nextMatch = matches.find(option => getIndexOfId(option.id) >= startIndex);
+      return nextMatch ?? matches[0];
+    }
+
+    return matches[0] ?? null;
+  };
+
+  /* Handle typed input */
+
+  // reset any typed value when an option is selected
+  baseState.selectOption = (ev, optionValue) => {
+    setValue(undefined);
+    selectOption(ev, optionValue);
+  };
+
+  // reset typed value when the input loses focus while collapsed, unless allowFreeform is true
+  const onBlur = (ev: React.FocusEvent<HTMLInputElement>) => {
+    if (!baseState.open && !allowFreeform) {
+      setValue(undefined);
+    }
+
+    triggerNativeProps.onBlur?.(ev);
+  };
+
+  baseState.setOpen = (ev, newState: boolean) => {
+    if (!newState && !allowFreeform) {
+      setValue(undefined);
+    }
+
+    setOpen(ev, newState);
+  };
+
+  // update value and active option based on input
+  const onChange = (ev: React.ChangeEvent<HTMLInputElement>) => {
+    const inputValue = ev.target.value;
+    // update uncontrolled value
+    baseState.setValue(inputValue);
+
+    // handle updating active option based on input
+    const matchingOption = getOptionFromInput(inputValue);
+    matchingOption && setActiveOption(matchingOption);
```

Comment:
nit: prefer `if` statements for conditionals like this
```suggestion
    if (matchingOption) {
      setActiveOption(matchingOption);
    }
```

Reply:
After chatting with Yvonne, I ended up always calling `setActiveOption` so that typing a non-matching string clears the active descendant :)